### DISCUSSION TITLE
is the forward protocol connection reusable for multiple messages?

### DISCUSSION THREAD

**Discussion by weltkante** [fluent/fluentd#3918] [Q&A] [answered]
I noticed the forward protocol specification saying that client can send multiple messages, yet the server is supposed to close the connection after the first message?

https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#network-transport-protocol

> 2. **Transport:** In this phase, clients send events to servers, and servers send response messages if client require it.
> 
> Once handshake phase completes, clients can use that connection in transport phase until disconnection.

https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#response

> Server SHOULD close the connection silently with no response when the `chunk` option is not sent.

This sounds like TCP connections are one time use unless you request acknowledgments? And even then the server is allowed to ignore the chunk option and close the connection anyways?

> Server MAY just ignore any options given.

**Comment by fujimotos** [comment, marked as answer]
> I noticed the forward protocol specification saying that client can send multiple messages, yet the server is supposed to close the connection after the first message?

This seems to be bad wording.

The forward protocol actually has `keepalive` flag  (see Handshake Message > HELO).
If this flag is set to true, the server will not close the connection after receiving a message.

Here are the corresponding lines in `in_foward.rb`:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_forward.rb#L225-L235

BTW, the `keepalive` flag is turned on by deafult:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_forward.rb#L48-L49

> I noticed the forward protocol specification saying that client can send multiple messages, yet the server is supposed to close the connection after the first message?

So yes, a client can send multiple messages using a single connection as
long as the server does not prohibit `keepalive`.
